File: src/falcon_core_old/math/domains/coupled_knob_domain.py
"""A collection of coupled knob domains to be attached together."""

from typing import TYPE_CHECKING

from .base_coupled_labelled_domain import BaseCoupledLabelledDomain
from .dependancies import Jsonable, Knob, Knobs, cast
from .labelled_domain import LabelledDomain
from .typing import Sequence

if TYPE_CHECKING:
    from .knob_domain import KnobDomain


class CoupledKnobDomain(BaseCoupledLabelledDomain[Knob], Jsonable):
    """A collection of coupled knob domains to be attached together."""

    def __init__(self, domains: list["KnobDomain"]):
        """Initialize the CoupledKnobDomain object."""
        labelled_domains = cast(Sequence[LabelledDomain[Knob]], domains)
        super().__init__(domains=labelled_domains)

    @property
    def knobs(self) -> "Knobs":
        """Return the knobs."""
        return Knobs(self.labels)

File: src/math/domains/CoupledKnobDomain.cpp
#include "falcon_core/math/domains/CoupledKnobDomain.hpp"

namespace falcon_core::math::domains {

CoupledKnobDomain::CoupledKnobDomain()
    : BaseCoupledLabelledDomain<instrument_interfaces::names::Knob>() {}
CoupledKnobDomain::CoupledKnobDomain(
    const std::vector<BaseLabelledDomainSP<instrument_interfaces::names::Knob>>&
        init)
    : BaseCoupledLabelledDomain<instrument_interfaces::names::Knob>(init) {}
const instrument_interfaces::names::KnobsSP CoupledKnobDomain::knobs() const {
  return std::make_shared<instrument_interfaces::names::Knobs>(labels());
}

}  // namespace falcon_core::math::domains

CEREAL_REGISTER_TYPE(falcon_core::math::domains::CoupledKnobDomain)
CEREAL_REGISTER_TYPE(falcon_core::math::domains::BaseCoupledLabelledDomain<
                     falcon_core::instrument_interfaces::names::Knob>)
CEREAL_REGISTER_POLYMORPHIC_RELATION(
    falcon_core::math::domains::BaseCoupledLabelledDomain<
        falcon_core::instrument_interfaces::names::Knob>,
    falcon_core::math::domains::CoupledKnobDomain)
